src/main.c

```diff
@@ -160,52 +160,66 @@ int is_ignored_class(const char *name) {
 void add_icon(Window w, int cmode)
 {
     struct TrayIcon *ti;
-    const char *classname = x11_get_window_class(tray_data.dpy, w);
+    const char *classname = NULL;
     /* Aviod adding duplicates */
     if ((ti = icon_list_find(w)) != NULL) {
         LOG_TRACE(("ignoring second request to embed 0x%x"
-                   "(requested cmode=%d, current cmode=%d\n",
+                   " (requested cmode=%d, current cmode=%d)\n",
             w, cmode, ti->cmode));
         return;
     }
     /* Dear Edsger W. Dijkstra, I see you behind my back =( */
-    if ((ti = icon_list_new(w, cmode)) == NULL) goto failed0;
+    if ((ti = icon_list_new(w, cmode)) == NULL) goto icon_allocation_failed;
     LOG_TRACE(("starting embedding for icon 0x%x, cmode=%d\n", w, cmode));
     x11_dump_win_info(tray_data.dpy, w);
 
+    classname = x11_get_window_class(tray_data.dpy, w);
+    if (classname == NULL) {
```

Comment:
Do I understand correctly that this check was missing and the code crashed later while checking the ignored list? Do you know why `x11_get_window_class()` returns NULL for your icon?

Reply:
Not exactly.

In the current version, `x11_get_window_class()` never returns `NULL`.
It is something that this patch is adding.

The problem is in `x11_get_window_class()` itself.
It does not check the `XGetClassHint()` return value.
`XGetClassHint()` may fail, in which case it does not update the `hint` output argument.
And as `hint` is a stack allocated structure that contains pointers, those pointers point to "random" addresses.

`XFree(hint.res_name)` causes a crash, as it is trying to deallocate at an address that is not a valid allocation.

---

I have no idea why `XGetClassHint()` fails.
It happens when I am running Google Chrome on Ubuntu, and I am trying to start an operation through a plugin, that uses USB HID to talk to my Ledger device.
I do not know if Chrome is trying to add a new icon or if it is doing something else.
But it is 100% reproducible.

So I can share, for example, the event details, if it would be of any help.
I have no understanding of how X11 works.
So, besides an obvious observation that error handing is missing, I do not understand the expectations of this interaction or anything of that sort.

I could probably run `stalonetray` with the debug output enabled, setting logging to the trace level.  This way, if the invalid `XGetClassHint()` call is producing an X11 error, I would be able to see it.

Reply:
Thanks. I missed that change in `x11_get_window_class()`. 

> I could probably run stalonetray with the debug output enabled, setting logging to the trace level. This way, if the invalid XGetClassHint() call is producing an X11 error, I would be able to see it.

That's usually an unavoidable step. Unfortunately, I have zero bandwidth for experimentation on my end, but I would be happy to review the log. I'm merging this change as it is useful regardless.